### ISSUE TITLE
TS2352:Type 'Object[keyof Object]' cannot be converted to type 'Date'

### ISSUE THREAD

**Issue by Karasuni** [Microsoft/TypeScript#14146]
**TypeScript Version:**  2.1.6

I've observed an issue with type casting in a function I've found and integrated from [here](https://gist.github.com/ghinda/8442a57f22099bdb2e34#gistcomment-1997213).

**Code**
```ts
  createFormData<Object>(object: Object, form?: FormData, namespace?: string): FormData {
    const formData = form || new FormData();
    for (let property in object) {
      if (!object.hasOwnProperty(property) || !object[property]) {
        continue;
      }
      const formKey = namespace ? `${namespace}[${property}]` : property;
      if (object[property] instanceof Date) {
        formData.append(formKey, (<Date>object[property]).toISOString());
//                                  ^
      } else if (typeof object[property] === 'object' && !(object[property] instanceof File)) {
        this.createFormData(object[property], formData, formKey);
      } else {
        formData.append(formKey, object[property]);
      }
    }
    return formData;
  }
```
Originally the `instanceof Date` switch had no typecasting and would result in a `TS2339:Property 'toISOString' does not exist on type 'Object[keyof Object]'. Since the object is already checked as an instance of `Date` I've added the cast.

**Expected behavior:**
It should not complain about typecasting.

**Actual behavior:**
`tsc` complains about not being able to cast an Object in an Object to a Date.

Current workaround: `(<Date><any>object[property]).toISOString());`

**Comment by jeffreymorlan** [comment]
```ts
      const p = object[property];
      if (p instanceof Date) {
        formData.append(formKey, p.toISOString());
      } else if (typeof p === 'object' && !(p instanceof File)) {
        this.createFormData(p, formData, formKey);
      } else {
        formData.append(formKey, p);
      }
```
